spec/integration/application/agent_spec.rb

```diff
@@ -0,0 +1,233 @@
+    Puppet[:hostprivkey] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1-key.pem')
+    Puppet[:hostcert] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1.pem')
+
+    facts = Puppet::Node::Facts.new(Puppet[:certname])
+    Puppet::Node::Facts.indirection.save(facts)
+  end
+
+  let(:server) { PuppetSpec::Puppetserver.new }
+  let(:agent) { Puppet::Application[:agent] }
+  let(:logfile) { tmpfile('agent.log') }
+  let(:args) { ['--test', '--logdest', logfile] }
+  let(:node) { Puppet::Node.new(Puppet[:certname], environment: 'production')}
+  let(:formatter) { Puppet::Network::FormatHandler.format(:rich_data_json) }
+
+  context 'server_list' do
+    it "uses the first server in the list" do
+      Puppet[:server_list] = '127.0.0.1'
+
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+    end
+
+    it "falls back, recording the first viable server in the report" do
+      server.start_server do |port|
+        Puppet[:server_list] = "puppet.example.com,#{Puppet[:server]}"
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+
+        report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+        expect(report.master_used).to eq("127.0.0.1:#{port}")
+      end
+    end
+
+    it "doesn't write a report if no servers could be contacted" do
+      Puppet[:server_list] = "puppet.example.com"
+
+      expect {
+        agent.command_line.args = args
+        agent.run
+      }.to exit_with(1)
+
+      # I'd expect puppet to update the last run report even if the server_list was
+      # exhausted, but it doesn't work that way currently, see PUP-6708
+      expect(File).to_not be_exist(Puppet[:lastrunreport])
+    end
+
+    it "omits master_used when not using server_list" do
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+
+      report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+      expect(report.master_used).to be_nil
+    end
+  end
+
+  context 'rich data' do
+    it "applies deferred values" do
+      catalog_handler = -> (req, res) {
+        catalog = compile_to_catalog(<<-MANIFEST, node)
+          notify { 'deferred':
+            message => Deferred('join', [[1,2,3], ':'])
+          }
+        MANIFEST
+
+        res.body = formatter.render(catalog)
+        res['Content-Type'] = formatter.mime
+      }
+
+      server.start_server(mounts: {catalog: catalog_handler}) do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(2)
```

Comment:
Is this the exit code of the puppet run? I'm not really sure what this is implying. Wouldn't it be better (or more clear) to ensure the notify message `1:2:3` is a part of the output of `agent.run`?

Reply:
The `exit_with` is necessary because the application calls `exit` which raises `SystemExit`. So we have to expect `exit_with` to prevent rspec from exiting. But yeah, I totally missed verifying that deferred writes the expected output. I'll fix that!

Reply:
Ah cool. Yeah, I saw the `exit_with` elsewhere, but I was confused we weren't validating anything else about the run. Thanks!

Reply:
Updated to expect the joined message.